File: library/src/main/java/com/view/jameson/library/CardAdapterHelper.java
package com.view.jameson.library;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import jameson.io.library.util.ScreenUtil;

/**
 * adapter中调用onCreateViewHolder, onBindViewHolder
 * Created by jameson on 9/1/16.
 */
public class CardAdapterHelper {
    private int mPagePadding = 25;
//    private int mShowLeftCardWidth = 15;
    private int mShowLeftCardWidth = 65;

    public void onCreateViewHolder(ViewGroup parent,  View itemView) {
        RecyclerView.LayoutParams lp = (RecyclerView.LayoutParams) itemView.getLayoutParams();
        lp.width = parent.getWidth() - ScreenUtil.dip2px(itemView.getContext(), 2 * (mPagePadding + mShowLeftCardWidth));
//        lp.width = ScreenUtil.dip2px(itemView.getContext(), 140);
//        System.out.println("测试，宽度"+lp.width+"/父类宽度"+parent.getWidth());
        itemView.setLayoutParams(lp);
    }

    public void onBindViewHolder(View itemView, final int position, int itemCount) {
        int padding = ScreenUtil.dip2px(itemView.getContext(), mPagePadding);
        itemView.setPadding(padding, 0, padding, 0);
        //原
        int leftMarin = position == 0 ? padding + ScreenUtil.dip2px(itemView.getContext(), mShowLeftCardWidth) : 0;
//        int leftMarin = position == 0 ? padding + ScreenUtil.dip2px(itemView.getContext(), 20) : 0;
//        int leftMarin =ScreenUtil.dip2px(itemView.getContext(), 20);
        //原
        int rightMarin = position == itemCount - 1 ? padding + ScreenUtil.dip2px(itemView.getContext(), mShowLeftCardWidth) : 0;
//        int rightMarin = position == itemCount - 1 ? padding + ScreenUtil.dip2px(itemView.getContext(), 20) : 0;
//        int rightMarin = ScreenUtil.dip2px(itemView.getContext(), 20);

//        System.out.println("测试左右边缘"+leftMarin +"/"+rightMarin);

        setViewMargin(itemView, leftMarin, 0, rightMarin, 0);
    }

    private void setViewMargin(View view, int left, int top, int right, int bottom) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        if (lp.leftMargin != left || lp.topMargin != top || lp.rightMargin != right || lp.bottomMargin != bottom) {
            lp.setMargins(left, top, right, bottom);
            view.setLayoutParams(lp);
        }
    }

    public void setPagePadding(int pagePadding) {
        mPagePadding = pagePadding;
    }

    public void setShowLeftCardWidth(int showLeftCardWidth) {
        mShowLeftCardWidth = showLeftCardWidth;
    }
}

File: library/src/main/java/com/view/jameson/library/CardAdapterHelper.kt
package com.view.jameson.library

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

import jameson.io.library.util.ScreenUtil

/**
 * adapter中调用onCreateViewHolder, onBindViewHolder
 *
 */
class CardAdapterHelper {
    private var mPagePadding = 25
    private var mShowLeftCardWidth = 65

    fun onCreateViewHolder(parent: ViewGroup, itemView: View) {
        val lp = itemView.layoutParams as RecyclerView.LayoutParams
        lp.width = parent.width - ScreenUtil.dip2px(itemView.context, (2 * (mPagePadding + mShowLeftCardWidth)).toFloat())
        itemView.layoutParams = lp
    }

    fun onBindViewHolder(itemView: View, position: Int, itemCount: Int) {
        val padding = ScreenUtil.dip2px(itemView.context, mPagePadding.toFloat())
        itemView.setPadding(padding, 0, padding, 0)
        val leftMarin = if (position == 0) padding + ScreenUtil.dip2px(itemView.context, mShowLeftCardWidth.toFloat()) else 0
        val rightMarin = if (position == itemCount - 1) padding + ScreenUtil.dip2px(itemView.context, mShowLeftCardWidth.toFloat()) else 0
        setViewMargin(itemView, leftMarin, 0, rightMarin, 0)
    }

    private fun setViewMargin(view: View, left: Int, top: Int, right: Int, bottom: Int) {
        val lp = view.layoutParams as ViewGroup.MarginLayoutParams
        if (lp.leftMargin != left || lp.topMargin != top || lp.rightMargin != right || lp.bottomMargin != bottom) {
            lp.setMargins(left, top, right, bottom)
            view.layoutParams = lp
        }
    }

    fun setPagePadding(pagePadding: Int) {
        mPagePadding = pagePadding
    }

    fun setShowLeftCardWidth(showLeftCardWidth: Int) {
        mShowLeftCardWidth = showLeftCardWidth
    }
}
